lib/ddtrace/correlation.rb

```diff
@@ -6,35 +6,109 @@ module Datadog
   # Contains behavior for managing correlations with tracing
   # e.g. Retrieve a correlation to the current trace for logging, etc.
   module Correlation
-    # Struct representing correlation
-    Identifier = Struct.new(:trace_id, :span_id, :env, :service, :version) do
-      def initialize(*args)
-        super
-        self.trace_id = trace_id || 0
-        self.span_id = span_id || 0
-        self.env = env || Datadog.configuration.env
-        self.service = service || Datadog.configuration.service
-        self.version = version || Datadog.configuration.version
+    # Represents current trace state with key identifiers
+    class Identifier
+      attr_reader \
+        :env,
+        :service,
+        :span_id,
+        :span_name,
+        :span_resource,
+        :span_service,
+        :span_type,
+        :trace_hostname,
+        :trace_id,
+        :trace_name,
+        :trace_origin,
+        :trace_process_id,
+        :trace_resource,
+        :trace_runtime_id,
+        :trace_sampling_priority,
+        :trace_service,
+        :version
+
+      # rubocop:disable Metrics/CyclomaticComplexity
+      # rubocop:disable Metrics/PerceivedComplexity
+      def initialize(
+        env: nil,
+        service: nil,
+        span_id: nil,
+        span_name: nil,
+        span_resource: nil,
+        span_service: nil,
+        span_type: nil,
+        trace_hostname: nil,
+        trace_id: nil,
+        trace_name: nil,
+        trace_origin: nil,
+        trace_process_id: nil,
+        trace_resource: nil,
+        trace_runtime_id: nil,
+        trace_sampling_priority: nil,
+        trace_service: nil,
+        version: nil
+      )
+        @env = env || Datadog.configuration.env
+        @service = service || Datadog.configuration.service
+        @span_id = span_id || 0
+        @span_name = span_name && span_name.dup.freeze
+        @span_resource = span_resource && span_resource.dup.freeze
+        @span_service = span_service && span_service.dup.freeze
+        @span_type = span_type && span_type.dup.freeze
+        @trace_hostname = trace_hostname && trace_hostname.dup.freeze
+        @trace_id = trace_id || 0
+        @trace_name = trace_name && trace_name.dup.freeze
+        @trace_origin = trace_origin && trace_origin.dup.freeze
+        @trace_process_id = trace_process_id
+        @trace_resource = trace_resource && trace_resource.dup.freeze
+        @trace_runtime_id = trace_runtime_id && trace_runtime_id.dup.freeze
+        @trace_sampling_priority = trace_sampling_priority
+        @trace_service = trace_service && trace_service.dup.freeze
+        @version = version || Datadog.configuration.version
```

Comment:
Hmmm, right. Still, I wonder if we could allow some of the mutation behavior but not all.

In particular, there's "you can provide a new name, resource, service, etc etc after the fact" and there's "you can actually mutate the existing `String` objects" after the fact.

Perhaps we could still allow users at any point to do

```ruby
span.foo = new_value
```

but not

```ruby
span.foo.concat("something")
```

The rule would be that everything we could inside the `Span` would be immutable, but not the `Span` itself. Would that work here?

Reply:
It's possible, but its odd, and difficult to communicate this to users. In the pipeline, they may have written code that uses `!` methods to mutate strings, so it's not intuitive if you can mutate the references but not the object.

One of the strategies I employed, then discarded, was rendering spans immutable, but when they passed through the pipeline, they'd be wrapped with a mutator class that would accept changes and build a new copy of the span if it was changed. I got rid of this out of concerns for performance: creating dups for every span that goes through the pipeline sounds like it could be expensive.

I'm inclined to leave this as is until we figure out what to do with span immutability. If we move our worker processes to Ractors, that might be a good time to introduce this, as the spans have to be made immutable anyways.